src/main/webapp/app/shared/server-date.service.ts

```diff
@@ -1,36 +1,89 @@
+    private recentClientDates = new Array<moment.Moment>();
+
+    constructor(private http: HttpClient) {}
+
+    /**
+     * get a new server date if necessary
+     */
+    updateTime(): void {
+        let shouldSync = false;
+        const now = moment(new Date());
+        if (this.recentClientDates.length > 4) {
+            this.recentClientDates.forEach((recentClientDate) => {
+                // only if all recent client dates (i.e. recent syncs are older than 60s)
+                if (now.diff(recentClientDate, 's') > 60) {
+                    shouldSync = true;
+                }
+            });
+        } else {
+            // definitely sync if we do not have 5 elements yet
+            shouldSync = true;
+        }
+        // TODO: one additional optimization could be to take the duration for request -> response into account here
+        if (shouldSync) {
+            // get new server date
+            this.http.get<string>(this.resourceUrl).subscribe((serverDate) => {
+                this.setServerDate(serverDate);
+            });
+        }
     }
 
     /**
-     * sets the current server date as unix
+     * adds the latest offset
      *
      * @param {string} date
      */
     setServerDate(date: string): void {
-        this.serverDate = moment(date).unix();
+        const serverDate = moment(date);
+        const clientDate = moment();
+        // save the most recent client date
+        this.recentClientDates.push(clientDate);
+        // calculate offset
+        const offset = serverDate.diff(clientDate, 'ms');
+        // save the most recent offset
+        this.recentOffsets.push(offset);
+        // remove oldest offset and client date if more than 5
+        if (this.recentOffsets.length > 5) {
+            this.recentOffsets.shift();
+            this.recentClientDates.shift();
+        }
+        // This would be faster and more reliable than using an interceptor with arbitrary REST calls
     }
 
     /**
      * returns the calculated current server date as moment
      */
-    now(): any {
-        // return moment.unix(this.serverDate);
-        // NOTE: quick fix which basically deactivates the calculation here until a better approach is implemented and tested
-        return moment();
+    now(): moment.Moment {
+        const clientDate = moment();
+        // return the client date if there are no offsets (e.g. when offline or before any api call was made)
+        if (this.recentOffsets.length === 0) {
+            return clientDate;
+        }
+        // take first offset if there are less than 5
+        let offset = this.recentOffsets[0];
+        // remove noise from offset if there are 5
+        if (this.recentOffsets.length === 5) {
+            const offsetsSorted = this.recentOffsets.sort((a, b) => b - a);
+            // remove lowest
+            offsetsSorted.shift();
+            // remove highest
+            offsetsSorted.pop();
```

Comment:
Array.sort in javascript is an in-place operation, which means this.recentOffsets would be modified to - also by the shift and pop operation.

Reply:
I changed it to work on a copy now